docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+#### Story U2
+
+A cluster admin would like to configure CAPI to remediate unhealthy MachinePool Machines to ensure a healthy fleet of Machines.
+
+#### Story U3
+
+A cluster admin updates a MachinePool to a newer Kubernetes version and would like to configure the strategy for that deployment so that the MachinePool will progressively roll out the new version of the machines. They would like this operation to cordon and drain each node to minimize workload disruptions.
+
+### Requirements
+
+--------------------------------------
+| ID | Requirement | Related Stories |
+|----|-------------|-----------------|
+| R1 | The MachinePool controller MUST create Machines representing the provider-specific resources in the MachinePool and enable cascading delete of infrastructure<br>machine resources upon delete of CAPI Machine. | U1 |
+| R2 | The machine health check controller MUST be able to select machines by label which belong to a MachinePool and remediate. | U2 |
+| R3 | The MachinePool API MUST provide an optional deployment strategy using the same type as MachineDeployment.Spec.Strategy. | U3 |
+| R4 | The Machine Controller MUST handle node drain for Machine Pool Machines with the same behavior as MachineDeployment Machines. | U1, U3 |
+
+### Implementation Details/Notes/Constraints
+
+- As an alternative, the `InfrastructureSelector` field could be attached to the CAPI MachinePool resource. Then the CAPI MachinePool controller could check it directly and it would be more clearly documented in code.
+
+- Since this is an optional feature, some providers will implement MachinePool Machines and others will not. While having MachinePool Machines represented improves the consistency of the UX when compared to MachineDeployments, they also create a split in the UX between different MachinePool providers: some will show Machines and others won't (the current behavior).
+
+  One alternative to that would be to create "hollow" MachinePool Machines, simply to have a representation of how many instances or replicas are in the MachinePool. These Machines would either be prevented from deletion, or would have no effect on actual provider resources when deleted. This doesn't seem like an obvious improvement over the previous option.
+
+  Since MachinePools are experimental and the several implementations are well-known, we expect that providers will implement MachinePool Machines fairly soon, and that the only implementations that cannot support it are AKSManagedMachinePool and AWSManagedMachinePool. The variation in UX is unfortunate but at least limited to these special types of MachinePool where it wraps a managed service. In those cases the behavior is likely to be understood by the user if they are familiar with the provider's service.
+
+### Risks and Mitigations
+
```

Comment:
Today the logic to convey replicas intent as Machines is provider agnostic and driven and abstracted by the core controllers. This approach inverts that and delegates the responsibility to each provider implementation which might result in a bad representation of intent as Machines CRs depending of the provider. May be mention this here.

Reply:
One goal of inverting the usual flow of creation here is to enable a MachinePool to be more responsive. A cloud provider's native scalable resource, such as an AWS AutoScaling Group or an Azure VM Scale Set, may remove and replace an instance without warning.

A provider's MachinePool controller can keep the infrastructure representation of the actual scalable resource more current, perhaps responding to native platform events. Then CAPI can sync its MachinePool Machines to match.

It's possible this creates temporary inconsistency in the overall representation of the MPMs. I would hope we can work that out cleanly in the implementations, but I will also make a note here.